### PR TITLE
bug 2062403 - Allow 'subcategories' longer than 29 characters

### PR DESCRIPTION
**Pull request by chutten**
### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry to `CHANGELOG.md` or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

File: glean_parser/schemas/metrics.2-0-0.schema.yaml [hunk 1]
```diff
@@ -21,7 +21,7 @@ definitions:
 
   dotted_snake_case:
     type: string
-    pattern: "^[a-z_][a-z0-9_]{0,29}(\\.[a-z_][a-z0-9_]{0,29})*$"
+    pattern: "^[a-z_][a-z0-9_]{1,40}(\\.[a-z_][a-z0-9_]{1,39})*$"
```
Comment (badboy) [inline glean_parser/schemas/metrics.2-0-0.schema.yaml]: Do we even need the upper limits here? It's correct I guess, but maybe overly complicated.
The total length is limited due to `maxLength`.
We want to express that: 
* Needs to start with a lower-case character (or `_` for some reason!)
* A dot is followed by an lower-case alphanumeric character (or `_` again!)

Your changed regex btw disallows a previously alowed category of `a.b`

File: tests/data/categories.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Any copyright is dedicated to the Public Domain.
+# https://creativecommons.org/publicdomain/zero/1.0/
+
+---
+$schema: moz://mozilla.org/schemas/glean/metrics/2-0-0
+
+we_used_to_have_subcategory_length:
```
Comment (badboy) [inline tests/data/categories.yaml:8]: This doesn't have any "subcategory", by which I think you mean dot-separated parts.

Maybe additionally this file should list the edge-cases I mentioned above (`a.b`), an actual "subcategory" longer than 29 characters, ...

**Review by badboy — CHANGES_REQUESTED** [review]

File: tests/data/categories.yaml (hunk 2 above)
Comment (badboy) [reply tests/data/categories.yaml:8]: Ah, I was somewhat wrong here. Of course this acts as _a single_ "subcategory" (hey, we never defined "subcategory"!) -- which now is longer. Still, I feel like we should cover the other cases.

File: tests/data/categories.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  but_we_dont_any_more:
+    type: counter
+    expires: never
+    description: A test metric
+    bugs: [https://bugzil.la/2062403]
+    data_reviews: [https://www.example.com]
+    notification_emails: [glean-team@mozilla.com]
```
Comment (badboy) [inline tests/data/categories.yaml:15]:
```suggestion
    notification_emails: [glean-team@mozilla.com]

# Total length: 39
# Subcategory: 29
start.abcdefghijklmnopqrstuvwxyzabc.end:
  but_we_dont_any_more: &defaults
    type: counter
    expires: never
    description: A test metric
    bugs: [https://bugzil.la/2062403]
    data_reviews: [https://www.example.com]
    notification_emails: [glean-team@mozilla.com]

# Total length: 40
# Subcategory: 30
start.abcdefghijklmnopqrstuvwxyzabcd.end:
  but_we_dont_any_more:
    <<: *defaults

a.b:
  but_we_dont_any_more:
    <<: *defaults
```

File: glean_parser/schemas/metrics.2-0-0.schema.yaml (hunk 1 above)
Comment (chutten) [reply glean_parser/schemas/metrics.2-0-0.schema.yaml]: Good points. Whether or not `a.b` is a good category name doesn't allow me to escape that this is a non-backwards-compatible change. Lemme move these groups to `*` so subcategories must all have at least one character (and, if so, it's not a digit) and the total length is only enforced by `maxLength`

File: tests/data/categories.yaml (hunk 2 above)
Comment (chutten) [reply tests/data/categories.yaml:8]: Fair 'nuff, this only exercises the first regex branch. I'll add an extra couple cases.

**Review by badboy — APPROVED** [review]
